pages/settings/wallets/index.js

```diff
@@ -1,29 +1,54 @@
 import { getGetServerSideProps } from '@/api/ssrApollo'
 import Layout from '@/components/layout'
 import styles from '@/styles/wallet.module.css'
-import { WalletCard } from '@/components/wallet-card'
-import { LightningAddressWalletCard } from './lightning-address'
-import { LNbitsCard } from './lnbits'
-import { NWCCard } from './nwc'
-import { LNDCard } from './lnd'
-import { CLNCard } from './cln'
-import { WALLETS } from '@/fragments/wallet'
-import { useQuery } from '@apollo/client'
-import PageLoading from '@/components/page-loading'
-import { LNCCard } from './lnc'
 import Link from 'next/link'
-import { Wallet as W } from '@/lib/constants'
+import { useWallets } from '@/components/wallet'
+import { useState } from 'react'
+import dynamic from 'next/dynamic'
 
-export const getServerSideProps = getGetServerSideProps({ query: WALLETS, authRequired: true })
+const WalletCard = dynamic(() => import('@/components/wallet-card'), { ssr: false })
+
+export const getServerSideProps = getGetServerSideProps({ authRequired: true })
 
 export default function Wallet ({ ssrData }) {
-  const { data } = useQuery(WALLETS)
+  const { wallets } = useWallets()
+
+  const [sourceIndex, setSourceIndex] = useState()
+  const [targetIndex, setTargetIndex] = useState()
+
+  const onDragStart = (i) => (e) => {
+    // e.dataTransfer.dropEffect = 'move'
+    // We can only use the DataTransfer API inside the drop event
+    // see https://html.spec.whatwg.org/multipage/dnd.html#security-risks-in-the-drag-and-drop-model
+    // e.dataTransfer.setData('text/plain', name)
+    // That's why we use React state instead
+    setSourceIndex(i)
+  }
+
+  const onDragEnter = (i) => (e) => {
+    setTargetIndex(i)
+  }
+
+  const onDragEnd = async (e) => {
+    setSourceIndex(null)
+    setTargetIndex(null)
+
+    if (sourceIndex === targetIndex) return
+
+    const newOrder = [...wallets]
 
-  if (!data && !ssrData) return <PageLoading />
-  const { wallets } = data || ssrData
-  const lnd = wallets.find(w => w.type === W.LND.type)
-  const lnaddr = wallets.find(w => w.type === W.LnAddr.type)
-  const cln = wallets.find(w => w.type === W.CLN.type)
+    const [source] = newOrder.splice(sourceIndex, 1)
+    const newTargetIndex = sourceIndex < targetIndex ? targetIndex - 1 : targetIndex
+    const append = sourceIndex < targetIndex
+
+    newOrder.splice(newTargetIndex + (append ? 1 : 0), 0, source)
+
+    await Promise.all(
+      newOrder.map((w, i) =>
+        w.setPriority(i).catch(console.error)
```

Comment:
This will trigger a mutation for every server wallet but only if the priority actually changed.

I forgot to add some error handling here. There is no feedback on errors so the order might get out of sync with the actual order in the server. Will add some.

Reply:
Add error toasts in f947a450. It will print separate errors for each call which looks bad but I think errors here should be rare:

![2024-07-08-131800_1920x1080_scrot](https://github.com/stackernews/stacker.news/assets/27162016/082a7ff9-d6b5-45a2-afa1-deecb94a3216)